corehq/apps/consumer_user/views.py

```diff
@@ -0,0 +1,195 @@
+from datetime import datetime, timedelta
+
+from django.contrib import messages
+from django.contrib.auth import logout
+from django.contrib.auth.forms import PasswordChangeForm
+from django.contrib.auth.models import User
+from django.core.signing import BadSignature, SignatureExpired, TimestampSigner
+from django.http import HttpResponse, HttpResponseRedirect
+from django.shortcuts import render
+from django.urls import reverse
+from django.utils.http import urlsafe_base64_decode
+from django.utils.translation import ugettext as _
+from django.views.decorators.debug import sensitive_post_parameters
+
+from two_factor.forms import AuthenticationTokenForm, BackupTokenForm
+from two_factor.views import LoginView
+
+from corehq.apps.consumer_user.models import (
+    ConsumerUser,
+    ConsumerUserCaseRelationship,
+    ConsumerUserInvitation,
+)
+from corehq.apps.domain.decorators import two_factor_exempt
+
+from ..users.models import CouchUser
+from .decorators import consumer_user_login_required
+from .forms.change_contact_details_form import ChangeContactDetailsForm
+from .forms.consumer_user_authentication_form import (
+    ConsumerUserAuthenticationForm,
+)
+from .forms.consumer_user_signup_form import ConsumerUserSignUpForm
+
+
+class ConsumerUserLoginView(LoginView):
+    form_list = (
+        ('auth', ConsumerUserAuthenticationForm),
+        ('token', AuthenticationTokenForm),
+        ('backup', BackupTokenForm),
+    )
+    invitation = None
+    hashed_invitation = None
+    template_name = 'p_login.html'
+
+    def get_form_kwargs(self, step=None):
+        """
+        Returns the keyword arguments for instantiating the form
+        (or formset) on the given step.
+        """
+        if self.invitation:
+            return {'invitation': self.invitation}
+        return {}
+
+    def get_success_url(self):
+        url = self.get_redirect_url()
+        if url:
+            return url
+        url = reverse('consumer_user:consumer_user_homepage')
+        return url
+
+    def get_context_data(self, **kwargs):
+        context = super(ConsumerUserLoginView, self).get_context_data(**kwargs)
+        if self.hashed_invitation:
+            extra_context = {}
+            this_is_not_me = reverse('consumer_user:consumer_user_register',
+                                     kwargs={
+                                         'invitation': self.hashed_invitation
+                                     })
+            extra_context['this_is_not_me'] = '%s%s' % (this_is_not_me,
+                                                        '?create_user=1')
+            context.update(extra_context)
+        context.update({'hide_menu': True})
+        return context
+
+
+def get_invitation_object_or_error(invitation):
```

Comment:
I was surprised that the two possible return values are completely different data types, and that "error" was actually an HttpResponse.

It might make it less surprising to the next person to be more explicit. Maybe ...

```suggestion
def get_invitation_object_or_error_response(invitation):
```

... or ...

```suggestion
def get_invitation_object_or_error(
    invitation: str,
) -> Union[ConsumerUserInvitation, HttpResponse]:
```
...

But, personally, I would use an exception, because an exception conveys the connotation/signifiance more than a normal return value does:

```python

class InvitationError(Exception):
    def __init__(self, msg, status, *args):
        self.msg = msg
        self.status = status
        super().__init__(*args)


class InvitationRedirect(InvitationError):
    def __init__(self, redirect_to):
        self.redirect_to = redirect_to
        super().__init__('', 302)


def get_invitation_obj(invitation):
    try:
        # ...
        if invitation_obj.accepted:
            url = reverse('consumer_user:consumer_user_login')
            raise InvitationRedirect(url)
        # ...
    except SignatureExpired:
        raise InvitationError(_("Invitation is expired"), status=400)


@two_factor_exempt
def register_view(request, invitation):
    try:
        invitation_obj = get_invitation_obj(invitation)
    except InvitationError as err:
        if hasattr(err, 'redirect_to')
            return HttpResponseRedirect(err.redirect_to)
        return HttpResponse(err.msg, status=err.status)
    #...

```

Reply:
(P.S. Nit: Considering `status` is always 400, and we never pass any other args to `InvitationError`, we could drop the `status` and `*args` params.)